### PR TITLE
report: map metrics to audits

### PR DESCRIPTION
**Pull request by paulirish**
Adds a little filter control so the audit list in Performance can be filtered by what's relevant to each metric.

➡️  See an interactive demo here: https://lighthouse-git-mapping-googlechrome.vercel.app/english/

<details>
<summary>The UX is ~inspired by the AMP page exp tool.</summary>

image
</details>

Addy's doc: go/cfspg
ref #11279

---------

image

-----------

### CODE REVIEW

**Comment by patrickhulce** [comment]
Sweeet! So glad we're starting down this path 🎉 

My primary concern with adding the filter without reworking the opportunity display is that it seems like it suggests the impact number is for that metric. Maybe it's mostly just a wording issue? "Filter be *relevant* metric:" would be a step in the right direction. Another quick alternative on my mind is removing the time estimate in the opportunity section when filtered. 

Other piece that stood out is the empty opportunities for most on TBT and CLS and how we can increase the prominence of diagnostics in those cases.

**Comment by paulirish** [comment]
I've given this a few updates. It's ready for a look. Description was updated with all the goodstuff.

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 1]
```diff
@@ -246,6 +250,13 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
         .filter(audit => audit.group === 'load-opportunities' && !Util.showAsPassed(audit.result))
         .sort((auditA, auditB) => this._getWastedMs(auditB) - this._getWastedMs(auditA));
 
+
+    const filterableMetrics = metricAudits.filter(a => !!a.relevantAudits);
+    // todo, only add if there are opps & diags rendered.
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:255]:
```suggestion
    // TODO: only add if there are opportunities & diagnostics rendered.
```

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 2]
```diff
@@ -298,6 +309,68 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
     element.appendChild(passedElem);
     return element;
   }
+
+  /**
+   * Render the control to filter the audits by metric. The filtering is done at runtime by CSS only
+   * @param {LH.ReportResult.AuditRef[]} filterableMetrics
+   * @param {HTMLDivElement} element
+   */
+  renderMetricAuditFilter(filterableMetrics, element) {
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:318]: Should this be in report ui features (like 3p filtering is)?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 3]
```diff
@@ -298,6 +309,68 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
+    filterableMetrics.unshift(/** @type {LH.ReportResult.AuditRef} */ ({acronym: 'All'}));
+    // thx https://codepen.io/surjithctly/pen/weEJvX
+    const filterBarEl = this.dom.createElement('div', 'lh-filterbar');
+    const filterBarSummaryEl = this.dom.createChildOf(filterBarEl, 'span', 'lh-filterbar__summary');
+    filterBarSummaryEl.textContent = 'Filter to relevant audits: ';
+    const labelSelectors = [];
+    const auditSelectors = [];
+    for (const metric of filterableMetrics) {
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:326]:
```suggestion
    for (const metric of [{acronym: 'All'}, ...filterableMetrics]) {
```

(don't mutate input parameters)

alt: `filterableMetric = [{blah}, ...filterableMetric]` up top

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 4]
```diff
@@ -298,6 +309,68 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
+        hidden: 'true',
+      });
+      const labelEl = this.dom.createChildOf(filterBarEl, 'label', 'lh-filterbar__label', {
+        for: elemId,
+      });
+      labelEl.textContent = metric.acronym || metric.id;
+      if (metric.acronym === 'All') {
+        radioEl.checked = true;
+      }
+      // Dynamically write some CSS, for the CSS-only filtering
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:342]: all this dynamic CSS seems much harder to follow than the equivalent JS would be. What reasons did you rule out JS?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 5]
```diff
@@ -298,6 +309,68 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
+    styleEl.textContent = `
+      ${labelSelectors.join(',\n')} {
+        background: var(--color-blue-A700);
+        color: var(--color-white);
+      }
+      /* If selecting non-All, hide all audits (and also the group header/description… */
+      .lh-filterbar__radio:checked:not(#metric-All) ~ .lh-audit-group .lh-audit,
+      .lh-filterbar__radio:checked:not(#metric-All) ~ .lh-audit-group .lh-audit-group__description,
+      .lh-filterbar__radio:checked:not(#metric-All) ~ .lh-audit-group .lh-audit-group__itemcount {
+        display: none;
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:364]: Are there alternatives to _hiding_ the filtered audits? I think keeping things in a constant place is beneficial. wdyt of reduce opacity?

File: lighthouse-core/report/html/report-styles.css [hunk 6]
```diff
@@ -537,6 +537,26 @@
   display: block;
 }
 
+
+.lh-filterbar {
```
Comment (connorjclark) [inline lighthouse-core/report/html/report-styles.css:541]: filterbar -> metricfilter or metricfilterbar or something. We have a 3p filter so "filter" is too ambiguous.

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 7]
```diff
@@ -118,6 +118,10 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
     if (fci) v5andv6metrics.push(fci);
     if (fmp) v5andv6metrics.push(fmp);
 
+
+    // TODO(paulirish): can drop these now that the acroynyms are in the config.
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js:122]: bump

File: types/lhr.d.ts [hunk 8]
```diff
@@ -87,6 +87,10 @@ declare global {
         weight: number;
         /** Optional grouping within the category. Matches the key of a Result.Group. */
         group?: string;
+        /** Any audit IDs closely relevant to this one */
```
Comment (connorjclark) [inline types/lhr.d.ts:90]: (applies to all comments)

please make a pass and 1) capitalize first letter 2) end with period

File: lighthouse-core/config/default-config.js [hunk 9]
```diff
@@ -416,17 +416,41 @@ const defaultConfig = {
...
+
+// go/lh-audit-metric-mapping
+const fcpRelevantAudits = [
+  // Roughly in order of impact
+  'server-response-time', 'render-blocking-resources', 'redirects', 'critical-request-chains',
+  'uses-text-compression', 'uses-rel-preconnect', 'uses-rel-preload', 'font-display', 'unminified-javascript', 'unminified-css',
+  'unused-css-rules',
+];
+
+defaultConfig = {
```
Comment (connorjclark) [inline lighthouse-core/config/default-config.js:429]: I don't understand why this was made mutable.

File: lighthouse-core/config/default-config.js [hunk 10]
```diff
@@ -416,17 +416,41 @@ const defaultConfig = {
       title: str_(UIStrings.bestPracticesGeneralGroupTitle),
     },
   },
+};
+
+// go/lh-audit-metric-mapping
+const fcpRelevantAudits = [
+  // Roughly in order of impact
```
Comment (connorjclark) [inline lighthouse-core/config/default-config.js:423]: is order used at all? What is impact? If none of this is used or well defined, I'd prefer just alphabetical order (and one per line :)

**Review by connorjclark — CHANGES_REQUESTED** [review]

**Comment by egsweeny** [comment]
Overall this is looking *amazing.* Just a quick fly-by suggestion re: verbiage for the copy 'Filter to relevant audits:'. It wasn't intuitive to me, but the current state might hold up if we don't like my alternatives (below).

Some alternatives:

1. Filter audits by metric 
2. Filter audits by applicable metric
3. Show audits relevant to: [alt: Filter audits relevant to:]
4. Show audits relevant to a specific metric:
5. Show audits to improve:

I'm currently leaning towards option 3, but am bike shedding and am flexible.

**Comment by connorjclark** [comment]
idea after talking to paul: would be nice to show metric adornments in the expanded audit view.

**Review by paulirish — COMMENTED** [review]
> idea after talking to paul: would be nice to show metric adornments in the expanded audit view.

i like this. I'm gonna do in a second PR as it adds more line changes then you'd think.

File: types/config.d.ts [hunk 11]
```diff
@@ -146,6 +146,8 @@ declare global {
         id: string;
         weight: number;
         group?: string;
+        relevantAudits?: string[];
```
Comment (connorjclark) [inline types/config.d.ts:149]: minor nit: swap order of these two new fields to match prototype field ids (and in below file)

**Review by connorjclark — APPROVED** [review]

File: lighthouse-core/lib/i18n/i18n.js [hunk 12]
```diff
@@ -41,6 +41,10 @@ const MESSAGE_I18N_ID_REGEX = / | [^\s]+$/;
     Intl.NumberFormat = IntlPolyfill.NumberFormat;
     Intl.DateTimeFormat = IntlPolyfill.DateTimeFormat;
   }
+  // Deal with buggy regex caching. https://github.com/andyearnshaw/Intl.js/issues/308
+  if (IntlPolyfill.__disableRegExpRestore) {
```
Comment (connorjclark) [inline lighthouse-core/lib/i18n/i18n.js:45]: big LOL

Comment (paulirish) [reply lighthouse-core/lib/i18n/i18n.js:45]: admittedly i should have googled before i build node 12 with `small-icu`. but yes. annoying.

**Pull request merged**